server/list.go

```diff
@@ -0,0 +1,239 @@
+package main
+
+import (
+	"fmt"
+)
+
+const (
+	// MyListKey is the key used to store the order of the owned todos
+	MyListKey = ""
+	// InListKey is the key used to store the order of received todos
+	InListKey = "_in"
+	// OutListKey is the key used to store the order of sent todos
+	OutListKey = "_out"
+)
+
+// ListStore represents the KVStore operations for lists
+type ListStore interface {
+	AddItem(item *Item) error
+	GetItem(itemID string) (*Item, error)
+	RemoveItem(itemID string) error
+
+	GetItemOrder(userID string, itemID string, listID string) (*OrderElement, int, error)
+	GetItemListAndOrder(userID string, itemID string) (string, *OrderElement, int)
+
+	Add(userID string, itemID string, listID string, foreignUserID string, foreignItemID string) error
+	Remove(userID string, itemID string, listID string) error
+	Pop(userID string, listID string) (*OrderElement, error)
+
+	GetList(userID string, listID string) ([]*OrderElement, error)
+}
+
+type listManager struct {
+	store       ListStore
+	getUserName func(string) string
+}
+
+// NewListManager creates a new listManager
+func NewListManager(store ListStore, getUserName func(string) string) *listManager {
+	return &listManager{
+		store:       store,
+		getUserName: getUserName,
+	}
+}
+
+func (l *listManager) Add(userID string, message string) error {
+	item := newItem(message)
+
+	l.store.AddItem(item)
+
+	if err := l.store.Add(userID, item.ID, MyListKey, "", ""); err != nil {
```

Comment:
Why are there two functions on the store for `Add`ing?
Another idea: maybe add comments to the functions -- that's where you can explain the differences (and the meaning of the parameters).

Reply:
This is something that is also buggering me. There are two different entities in place here.
First one would be the items. Items are invariable during its lifetime, and just record basic information about the todo: ID, Create time, and Message.
On the other hand, there are the "lists". Lists are stored as a list of order elements. On the previous implementation, this order elements were just the IDs of the items. Now the order elements have more information.
The AddItem is used to store items.
The Add is used to add a new Order Element to a list.
I will add comments to make it more understandable, but maybe there should be a bigger change here. If you have any idea, please let me know.

Reply:
I suppose the best we can do for now is hide this complexity in the manager, and not let it leak into the command runner or the store. This is the "business-logic". 

We should leave it for another PR to refactor how lists and items are stored to simplify this (if that's possible).